src/main/kotlin/de/bund/digitalservice/useid/credentials/CredentialController.kt

```diff
@@ -0,0 +1,87 @@
+package de.bund.digitalservice.useid.credentials
+
+import de.bund.digitalservice.useid.events.AuthenticateEvent
+import de.bund.digitalservice.useid.events.EventService
+import de.bund.digitalservice.useid.events.EventType
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
+import org.springframework.http.HttpStatus
+import org.springframework.http.MediaType
+import org.springframework.http.ResponseEntity
+import org.springframework.web.bind.annotation.PathVariable
+import org.springframework.web.bind.annotation.PostMapping
+import org.springframework.web.bind.annotation.PutMapping
+import org.springframework.web.bind.annotation.RequestBody
+import org.springframework.web.bind.annotation.RestController
+import java.util.UUID
+
+internal const val CREDENTIALS_BASE_PATH = "/api/v1/credentials"
+
+@RestController
+@ConditionalOnProperty(name = ["features.desktop-solution-enabled"], havingValue = "true")
+class CredentialController(
+    private val credentialService: CredentialService,
+    private val eventService: EventService,
+) {
+    @PostMapping(path = [CREDENTIALS_BASE_PATH])
+    fun startRegistration(
+        @RequestBody startRegistrationRequest: StartRegistrationRequest,
+    ): ResponseEntity<Any> {
+        val credential = credentialService.startRegistration(
+            startRegistrationRequest.widgetSessionId,
+            startRegistrationRequest.refreshAddress,
+        )
+
+        return ResponseEntity
+            .status(HttpStatus.ACCEPTED)
+            .contentType(MediaType.APPLICATION_JSON)
+            .body(
+                StartRegistrationResponse(
+                    credential.credentialId,
+                    credential.pckCreationOptions.toCredentialsCreateJson(),
+                ),
+            )
+    }
+
+    @PutMapping(
+        path = ["$CREDENTIALS_BASE_PATH/{credentialId}"],
+        headers = ["Content-Type=application/json"],
+    )
+    fun completeRegistration(
+        @PathVariable credentialId: UUID,
+        @RequestBody publicKeyCredentialJson: String,
+    ): ResponseEntity<Any> {
+        credentialService.finishRegistration(credentialId, publicKeyCredentialJson)
+
+        val credential = credentialService.startAuthentication(credentialId)
+        sendAuthenticateEventToWidget(credential)
+
+        return ResponseEntity
+            .status(HttpStatus.ACCEPTED)
+            .contentType(MediaType.APPLICATION_JSON)
+            .build()
+    }
+
+    private fun sendAuthenticateEventToWidget(credential: Credential) {
+        eventService.publish(
+            AuthenticateEvent(credential.credentialId, credential.assertionRequest!!.toCredentialsGetJson()),
```

Comment:
I don't have a suggestion, but maybe we could get rid of the double bang operator.

Reply:
Added a null check before, which throws a more descriptive exception.